flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBOrderedStore.java

```diff
@@ -0,0 +1,283 @@
+		this.batchWrapper = batchWrapper;
+		this.groupPrefixBytes = createKeyGroupBytes(keyGroupId);
+	}
+
+	private byte[] createKeyGroupBytes(int keyGroupId) {
+
+		outputStream.reset();
+
+		try {
+			outputView.writeShort(keyGroupId);
+		} catch (IOException e) {
+			throw new FlinkRuntimeException("Could not write key-group bytes.", e);
+		}
+
+		return outputStream.toByteArray();
+	}
+
+	@Override
+	public void add(@Nonnull T element) {
+		byte[] elementBytes = serializeElement(element);
+		try {
+			batchWrapper.put(columnFamilyHandle, elementBytes, DUMMY_BYTES);
+		} catch (RocksDBException e) {
+			throw new FlinkRuntimeException("Error while getting element from RocksDB.", e);
+		}
+	}
+
+	@Override
+	public void remove(@Nonnull T element) {
+		byte[] elementBytes = serializeElement(element);
+		try {
+			batchWrapper.remove(columnFamilyHandle, elementBytes);
+		} catch (RocksDBException e) {
+			throw new FlinkRuntimeException("Error while removing element from RocksDB.", e);
+		}
+	}
+
+	/**
+	 * This implementation comes at a relatively high cost per invocation. It should not be called repeatedly when it is
+	 * clear that the value did not change. Currently this is only truly used to realize certain higher-level tests.
+	 *
+	 * @see org.apache.flink.runtime.state.heap.CachingInternalPriorityQueueSet.OrderedSetStore
+	 */
+	@Override
+	public int size() {
+
+		int count = 0;
+		try (final RocksToJavaIteratorAdapter iterator = orderedIterator()) {
+			while (iterator.hasNext()) {
+				iterator.next();
+				++count;
+			}
+		}
+
+		return count;
+	}
+
+	@Nonnull
+	@Override
+	public RocksToJavaIteratorAdapter orderedIterator() {
+
+		flushWriteBatch();
+
+		return new RocksToJavaIteratorAdapter(
+			new RocksIteratorWrapper(
+				db.newIterator(columnFamilyHandle, readOptions)));
+	}
+
+	/**
+	 * Ensures that recent writes are flushed and reflect in the RocksDB instance.
+	 */
+	private void flushWriteBatch() {
+		try {
+			batchWrapper.flush();
+		} catch (RocksDBException e) {
+			throw new FlinkRuntimeException(e);
+		}
+	}
+
+	private static boolean isPrefixWith(byte[] bytes, byte[] prefixBytes) {
+		for (int i = 0; i < prefixBytes.length; ++i) {
+			if (bytes[i] != prefixBytes[i]) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	private byte[] serializeElement(T element) {
+		try {
+			outputStream.reset();
+			outputView.writeShort(keyGroupId);
+			byteOrderProducingSerializer.serialize(element, outputView);
+			return outputStream.toByteArray();
+		} catch (IOException e) {
+			throw new FlinkRuntimeException("Error while serializing the element.", e);
+		}
+	}
+
+	private T deserializeElement(byte[] bytes) {
+		try {
+			// TODO introduce a stream in which we can change the internal byte[] to avoid creating instances per call
+			ByteArrayInputStreamWithPos inputStream = new ByteArrayInputStreamWithPos(bytes);
+			DataInputViewStreamWrapper inputView = new DataInputViewStreamWrapper(inputStream);
+			inputView.readShort();
```

Comment:
Maybe we could use `inputView.skip(2)` here.

Reply:
`skip()` also has a return value that should be checked (in general, maybe not for this particular stream that we use), similar to what `skipBytes(...)`. I think still doing the check is cleaner and then I think we can also use `readShort`, and it is easier to identify as the counterpart of the corresponding `writeShort(...)`.